### ISSUE TITLE
custom error messages

### ISSUE THREAD

**Issue by GochoMugo** [confuser/node-recaptcha-middleware#2]
First, thanks for your contribution to the OSS community. I have just started using this module in an application that has already been pushed to production. It's in my best interest to see this module flourish in its use-case.

I have you given any thought of adding custom error messages. Typically error messages, passed by middleware throught `next()`, are shown to the user in error messages. It would be nice to have the ability to customize this error messages without so much of `regex` and such.

I would be willing to work on a PR but I would like to get your feedback first.

**Comment by confuser** [comment]
Thank you for opening an issue first rather than just sending an immediate pull request 👍 

Which error messages are you referring to? I wrote this module quite defensively, so some of them will be quite rare/should never occur e.g. `'Recaptcha not successful but no error codes provided'`

**Comment by GochoMugo** [comment]
I do think that the `Missing g-recaptcha-response field` message, in particular, is a bit assumptive that the end-user understands technical workings of captcha and form processing, in general. It would be nice to add some comments/advice to the user on how to handle the error.

Or rather, we could add an error code that can be referred to later in the middleware stack to suit more-specific purposes.

How do you handle it?

**Comment by GochoMugo** [comment]
Sorry! Closed it by accident. Always tend to click on the less-shiny button! My bad!

**Comment by confuser** [comment]
That particular check is there as a defence mechanism to avoid crashing the process. Ideally that should be handled at a validation layer, before this middleware is called, further up the chain; this would allow for complete control over what behaviour your system should exhibit in such circumstances.

**Comment by GochoMugo** [comment]
I would rather have this middleware stacked above validation layer (in the general sense). The main purpose of ReCaptcha is stop bots. By having the middleware hit as early as possible, we can avoid any CPU cycles and time being spent on requests that may eventually be flagged as spam.

**Comment by confuser** [comment]
Fair enough. Then I propose a interface change (v2) for this to become more accommodating.

``` js
var opts =
  { publicKey: ''
  , secretKey: ''
  , errors:
    { validation: new Error('Missing g-recaptcha-response field')
    , missingBody: new Error('Missing body response from recaptcha')
    , missingError: new Error('Recaptcha not successful but no error codes provided')
    , recaptchaErrorHandler: function (errors) {
        return new Error(errors.join(', '))
      }
    } 
  }
middleware(opts)
```

Default options above. I'll see if I get some time during lunch one day this week to update the module.

**Comment by GochoMugo** [comment]
How about using error codes. It will highly lessen the work and code to write to support it. All we need to add would be something like,

``` js
var error = new Error('Missing g-recaptcha-response field')
error.code = 'ECAPTCHANOFIELD'
```

In the later middlewares, should one want to do something custom, they could do,

``` js
app.use(function (err, req, res, next) {
  if (err.code !== 'ECAPTCHANOFIELD') return next(err)

  res.render('register', {
    body: req.body,
    error: {
      message: "Prove you are Human by solving the ReCaptcha shown!",
    },
  })
})
```

It would not break the API and way simpler and easier to maintain. The current error messages remain, yes, but if the programmer wants to use different error messages, they have the control to do so using the codes.

I have seen this also done in [expressjs/csurf](https://github.com/expressjs/csurf#custom-error-handling)'s custom error handling. I think it is a nice approach.

**Comment by confuser** [comment]
Whilst error code will suffice in your circumstances, it may not for others. I'd rather make it flexible to cater for as many cases as possible. This module shouldn't really require any further maintenance after that other than dependency updates.

**Comment by GochoMugo** [comment]
Could you give an example of such a circumstance, so I could understand the consideration better?

**Comment by confuser** [comment]
Sure. Some applications use custom errors (not just codes) to handle different logic at in their error handlers. By allowing them to define the value it it's entirety, they can use whatever they like, without having to do any additional if checks (i.e. mapping an error code to their custom Error objects etc).

**Comment by GochoMugo** [comment]
I agree. :+1: 

What do you intend to have the option `recaptchaErrorHandler` handle?

**Comment by confuser** [comment]
Currently any errors passed back from recaptcha itself are joined into a single error message. Turning it into a function allows any user to change it into something more meaningful should they need it.

**Comment by GochoMugo** [comment]
Good call.

**Comment by confuser** [comment]
Published as `2.0.0`
